ui/emojis: New module for inserting/displaying emojis

akirabaruah opened the pull request:
Uses the emojify package to enable displaying and inserting emojis. Also adds a new binding "SPC i e" to insert an Emoji.

AlexAnarcho commented:
Would be a nice feature, i m struggeling with emojis in doom emacs

hlissner commented on modules/ui/emojis/config.el:
```diff
@@ -0,0 +1,13 @@
+                        (if (featurep! +unicode) 'unicode))))
+
+(use-package! emojify
+  :defer t
```
Remove `:defer t`. It's redundant with `:hook`.

akirabaruah replied:
Done!

hlissner commented on modules/ui/emojis/config.el:
```diff
@@ -0,0 +1,13 @@
+
+(use-package! emojify
+  :defer t
+  :hook (after-init . global-emojify-mode)
```
Can this be deferred later? Perhaps until `doom-first-buffer-hook`? Or at most `doom-first-input-hook`?

akirabaruah replied:
Yep, `doom-first-buffer-hook` seems more appropriate.

hlissner commented on modules/ui/emojis/config.el:
```diff
@@ -0,0 +1,13 @@
+;;; ui/emojis/config.el -*- lexical-binding: t; -*-
+
+(setq emojify-styles
+      (seq-filter #'identity
+                  (list (if (featurep! +ascii) 'ascii)
+                        (if (featurep! +github) 'github)
+                        (if (featurep! +unicode) 'unicode))))
```
This is emojify configuration, it should be in a `:config` of the `emojify` use-package block.

akirabaruah replied:
Done!

hlissner commented on modules/ui/emojis/config.el:
```diff
@@ -0,0 +1,13 @@
+;;; ui/emojis/config.el -*- lexical-binding: t; -*-
+
+(setq emojify-styles
+      (seq-filter #'identity
```
Use `cl-remove-if-not`. Doom loads `cl-lib` by default, but not `seq` (at least, not in interactive sessions), so there is no hard guarantee `seq-filter` will be available.

akirabaruah replied:
Good to know. Fixed!

hlissner commented on modules/ui/emojis/packages.el:
```diff
@@ -0,0 +1,6 @@
+;;; ui/emojis/packages.el
+
+(package! emojify
+  :recipe (:host github :repo "iqbalansari/emacs-emojify")
```
Why the explicit recipe? `emojify` appears to be on MELPA and it appears to be from the same repo.

akirabaruah replied:
No reason besides ignorance - thanks for pointing it out!

hlissner commented on modules/ui/emojis/config.el:
```diff
@@ -0,0 +1,11 @@
+  :hook (doom-first-buffer . global-emojify-mode)
+  :config
+  (setq emojify-styles
+        (cl-remove-if-not #'identity
+                          (list (if (featurep! +ascii) 'ascii)
+                                (if (featurep! +github) 'github)
+                                (if (featurep! +unicode) 'unicode))))
```
Actually, could you replace this with:

```elisp
(setq emojify-styles
      (delq
       nil (list (if (featurep! +ascii) 'ascii)
                 (if (featurep! +github) 'github)
                 (if (featurep! +unicode) 'unicode))))
```

akirabaruah replied:
Ooh that's nice and concise. Done!

hlissner reviewed (changes requested):
Looks good! Two things occurred to me:

1. Could you add an entry for this module in docs/modules.org. e.g.

   ```diff
    + [[file:../modules/ui/doom/README.org][doom]] - TODO
    + [[file:../modules/ui/doom-dashboard/README.org][doom-dashboard]] - TODO
    + [[file:../modules/ui/doom-quit/README.org][doom-quit]] - TODO
   ++ emoji =+ascii +github +unicode= - Adds emoji support to Emacs
    + fill-column - TODO
    + [[file:../modules/ui/hl-todo/README.org][hl-todo]] - TODO
    + [[file:../modules/ui/hydra/README.org][hydra]] - TODO
   ```

2. Could you rename the module `:ui emoji`? There's no consensus on whether emoji plural is "emoji" or "emojis", but for the sake of keystrokes I prefer the former.

akirabaruah commented:
1. Added `emoji` entry in `docs/modules.org`
2. Renamed module to `:ui emoji` :+1: yeah plural's awkward :sweat_smile: I speak Japanese (where "emoji" is both singular/plural), but I still find myself saying "emojis" in English :shrug:

hlissner commented:
Looks good to me. Thanks for your help!

agsdot commented:
This is a pretty awesome feature add to Doom. Really appreciate this PR coming in.

akirabaruah commented:
@agsdot glad you found it useful 😄 

Real hats off to @iqbalansari (author of the package) and @hlissner for this awesome project!

agsdot commented:
Yeah, props to all of you though @akirabaruah , @hlissner and @iqbalansari. 

Before this package to generate emoji's I had to either be on my mobile device and keyboard or on a desktop and use https://emojipedia.org/. This package and nice keybinding of SPC i e, lends a very nice and rapid way to find emoji when typing.

Thanks once again!
